gpt4all-bindings/typescript/src/util.js

```diff
@@ -4,12 +4,35 @@ const { performance } = require("node:perf_hooks");
 const path = require("node:path");
 const { mkdirp } = require("mkdirp");
 const md5File = require("md5-file");
-const { DEFAULT_DIRECTORY, DEFAULT_MODEL_CONFIG } = require("./config.js");
+const {
+    DEFAULT_DIRECTORY,
+    DEFAULT_MODEL_CONFIG,
+    DEFAULT_MODEL_LIST_URL,
+} = require("./config.js");
+
+async function listModels(
+    options = {
+        url: DEFAULT_MODEL_LIST_URL,
+    }
+) {
+    if (options.file && existsSync(options.file)) {
+        console.debug("exists", options.file);
+        const fileContents = await fsp.readFile(options.file, "utf-8");
+        const modelList = JSON.parse(fileContents);
+        return modelList;
+    } else if (options.url) {
+        const res = await fetch(options.url);
+
+        if (!res.ok) {
+            throw Error(
+                `Failed to retrieve model list from ${url} - ${res.status} ${res.statusText}`
+            );
+        }
+        const modelList = await res.json();
+        return modelList;
+    }
 
-async function listModels() {
-    const res = await fetch("https://gpt4all.io/models/models.json");
-    const modelList = await res.json();
-    return modelList;
+    return [];
```

Comment:
Should this be intended behavior? 🤔  I think we should throw if options nor url was provided

Reply:
yea not great. added a check for missing source option added a test checking for that behavior